contracts/PeeranhaCommunityToken.sol

```diff
@@ -0,0 +1,138 @@
+
+  event AddBalance(uint256 indexed amount);   // name
+
+  constructor(address tokenAddress, uint256 maxRewardPerPeriod, uint256 activeUsersInPeriod, address peeranhaCommunityTokenFactoryAddress) {
+    if (tokenAddress == address(0))
+      communityTokenContainer.info.tokenAddress = address(0x0000000000000000000000000000000000001010); // chack Address !
+    else
+      communityTokenContainer.info.tokenAddress = tokenAddress;
+    communityTokenContainer.info.name = IERC20MetadataUpgradeable(communityTokenContainer.info.tokenAddress).name();
+    communityTokenContainer.info.symbol = IERC20MetadataUpgradeable(communityTokenContainer.info.tokenAddress).symbol();
+    communityTokenContainer.info.maxRewardPerPeriod = maxRewardPerPeriod;
+    communityTokenContainer.info.activeUsersInPeriod = activeUsersInPeriod;
+    communityTokenContainer.info.maxRewardPerUser = maxRewardPerPeriod / activeUsersInPeriod;
+    communityTokenContainer.info.createTime = CommonLib.getTimestamp();
+    communityTokenContainer.info.peeranhaCommunityTokenFactoryAddress = peeranhaCommunityTokenFactoryAddress;
+  }
+
+  // This is to support Native meta transactions
+  // never use msg.sender directly, use _msgSender() instead
+  function _msgSender()
+      internal
+      override
+      view
+      returns (address sender)
+  {
+    return NativeMetaTransaction._msgSender();
+  }
+
+  function addBalance(uint256 amount) external {  // name
+    address owner = _msgSender();
+    IERC20Upgradeable(communityTokenContainer.info.tokenAddress).transferFrom(owner, address(this), amount);
+
+    emit AddBalance(amount);
+  }
+
+  function updateCommunityRewardSettings(uint256 maxRewardPerPeriod, uint256 activeUsersInPeriod) external override {
+    communityTokenContainer.info.maxRewardPerPeriod = maxRewardPerPeriod;
+    communityTokenContainer.info.activeUsersInPeriod = activeUsersInPeriod;
+    communityTokenContainer.info.maxRewardPerUser = maxRewardPerPeriod / activeUsersInPeriod;
+  }
+
+  function getBalance() public view override returns(uint256) {   // public?
+    uint256 balance = IERC20Upgradeable(communityTokenContainer.info.tokenAddress).balanceOf(address(this));
+    require(balance >= communityTokenContainer.info.frezeTokens, "error_balance"); // todo: test??
+
+    return balance - communityTokenContainer.info.frezeTokens;
+  }
+
+  // get pool
+  function getTotalPeriodReward(uint16 period) private view returns(uint256) {
+    return communityTokenContainer.periodPool[period];
+  }
+
+  // set pool
+  function setTotalPeriodReward(RewardLib.PeriodRewardShares memory periodRewardShares, uint16 period) external override {
+    require(_msgSender() == communityTokenContainer.info.peeranhaCommunityTokenFactoryAddress, "only_community_token_factory_can_call_this_action");
+
+    uint256 totalPeriodReward = communityTokenContainer.info.maxRewardPerPeriod * 10 ** IERC20MetadataUpgradeable(communityTokenContainer.info.tokenAddress).decimals();
```

Comment:
why do we need to add `* 10 ** IERC20MetadataUpgradeable(communityTokenContainer.info.tokenAddress).decimals()`?
Can we just set `maxRewardPerPeriod` already multiplied by that value? I think it is something that is typically expected.